sympy/polys/rings.py

```diff
@@ -2531,20 +2531,318 @@ def compose(f, x, a=None):
+        `q` and `r`, the quotient and remainder, respectively, satisfying `f = gq + r`,
+        where `deg(r) < deg(g)`.
+
+        In the context of the `prem` method, multivariate polynomials in a ring
+        like `R[x,y,z]` are treated as univariate polynomials with coefficients
+        that are polynomials, such as `R[x,y][z]`. The pseudo-quotient `q` and
+        pseudo-remainder `r` with respect to `z` when dividing `f` by `g`
+        satisfy `mf = gq + r`, where `deg(r,z) < deg(g,z)` and
+        `m = LC(g,z)**(deg(f,z) - deg(g,z)+1)`.
+
+        The `prem` method returns the pseudo-remainder `r`,
+        the `pquo` method returns the pseudo-quotient `q`,
+        and `pdiv` returns the tuple `(q, r)`.
+
+        Parameters
+        ==========
+
+        g : PolyElement
+            The polynomial to divide `self` by.
+        x : generator or generator index, optional
+            The main variable of the polynomials and default is first generator.
+
+        Returns
+        =======
+
+        PolyElement
+            The pseudo-remainder polynomial.
 
-    def pdiv(f, g):
-        return f.ring.dmp_pdiv(f, g)
+        Raises
+        ======
 
-    def prem(f, g):
-        return f.ring.dmp_prem(f, g)
+        ZeroDivisionError : If `g` is the zero polynomial.
+
+        Examples
+        ========
+
+        >>> from sympy.polys import ring, ZZ
+        >>> R, x, y = ring("x, y", ZZ)
+        >>> f = x**2 + x*y
+        >>> g = 2*x + 2
+        >>> f.prem(g) # first generator is chosen by default if it is not given
+        -4*y + 4
+        >>> f.prem(g, y) # generator is given
+        0
+        >>> f.prem(g, 1) # generator index is given
+        0
+
+        See Also
+        ========
+
+        pquo
+        pdiv
+        pexquo
+
+        """
+        f = self
+        x = f.ring.index(x)
+        df = f.degree(x)
+        dg = g.degree(x)
+
+        if dg < 0:
+            raise ZeroDivisionError('polynomial division')
+
+        r, dr = f, df
+
+        if df < dg:
+            return r
+
+        N = df - dg + 1
+
+        lc_g = g.coeff_wrt(x, dg)
+
+        xp = f.ring.gens[x]
+
+        while True:
+
+            lc_r = r.coeff_wrt(x, dr)
+            j, N = dr - dg, N - 1
+
+            R = r * lc_g
+            G = g * lc_r * xp**j
+            r = R - G
+
+            dr = r.degree(x)
+
+            if dr < dg:
+                break
+
+        c = lc_g ** N
+
+        return r * c
+
+    def pdiv(self, g, x=None):
+        """
+        Computes the pseudo-division of the polynomial ``self`` with respect to ``g``.
+
+        The pseudo-division algorithm finds the pseudo-remainder `r` such that the relationship
+        `ma = bq + r` is satisfied, where `m` is the multiplier and `q` is the pseudo-quotient.
+        `r` and `q` are polynomials in `x`, with `degree(r, x) < degree(b, x)`. The multiplier
+        `m` is defined as `lcoeff(b, x) ^ (degree(a, x) - degree(b, x) + 1)` and is free of `x`.
```

Comment:
You are referring here to `a` and `b` but there is no `a` and `b` anywhere else.

Reply:
Also this looks like it is more or less copied from the Maple docs. We need to be careful about plagiarism and copyright so don't copy from there.
https://www.maplesoft.com/support/help/maple/view.aspx?path=prem

Reply:
> Also this looks like it is more or less copied from the Maple docs. We need to be careful about plagiarism and copyright so don't copy from there. https://www.maplesoft.com/support/help/maple/view.aspx?path=prem

Sorry, my apologies I will remember and careful about these from the next time.